### DISCUSSION TITLE
Translations - maintenance and automation

### DISCUSSION THREAD

**Discussion by amezin** [ddterm/gnome-shell-extension-ddterm#567] [General]
So initially I didn't want to add translations - because I expect that almost everyone who uses terminal is able to read English (at least, typical UI labels in English)

But at first people started requesting translations, and then went adding them in pull requests - and who am I to deny contributions?

But it adds significant maintenance burden.

So, in this discussion, please share your ideas, if you have any:

- How to make the maintenance process for translations easier? By this I mean: ensure that translations are at least somewhat complete before shipping them. Ensure that newly added strings are translated.

- How to make translation process easier? For example, good web-based tools for .po files

I choose Weblate as the tool because it creates commits with authorship information - so translators get added as contributors on GitHub, which seemed important to me. And, also, Weblate adoption added a few new translations almost immediately.

But I'm open to any suggestions.

@ddterm/translators

**Comment by amezin** [comment]
One thing I don't like about Weblate though - when there's a merge conflict, you have to merge all changes from Weblate to resolve it (and until you do that, all work is stopped). Happened again today - so I had to merge empty translations into the repository.

**Comment by daPhipz** [comment]
> ensure that translations are at least somewhat complete before shipping them

If you want to ship translations only when they are mostly complete, I also would advise for direct contributions via PRs. That way, you can control when a translation gets merged into `master` which will end up in the release. Or do you have to manually merge the Weblate contributions as well? If so, are those Weblate PRs divided per language, or is it one PR to update all translations?

> Ensure that newly added strings are translated

This is up to the code contributors/PR reviewers.

> How to make translation process easier? For example, good web-based tools for .po files

I think that (at least for somewhat experienced translators) "raw" editing of the `po` files is the best way to edit translations - be it in a text editor or via a GUI app like Poedit or Gtranslator. At lest, that's my preferred workflow. I haven't used Weblate this much, to be honest, but I don't like that they only show one string at a time. Sometimes, the context with
 other strings is important to get some more information for the translation, this is missing then.
However, I see that translating software is a good way for not-so-tech-savy people to contribute and give back to the FOSS community. Hence, the barrier of entry should be quite low. Learning git, GitHub etc. might already be too much of a hassle...
What do you think?

**Reply by amezin** [reply]
> Or do you have to manually merge the Weblate contributions as well? If so, are those Weblate PRs divided per language, or is it one PR to update all translations?

Weblate creates a pull request but, unfortunately, only one for all languages. It squashes all commits by author, though, so I can cherry-pick them...

> > Ensure that newly added strings are translated

> This is up to the code contributors/PR reviewers.

Hm. So imagine I'm adding a new UI element. What should I do? Asking all translators to translate it immediately, before merging the PR, is not an option, I think.

I started to think about an issue generator using GitHub Actions. Check the translation files, and if some strings aren't translated, raise/update an issue, one issue per language. But I'm not sure when I'll be able to work on it.

In theory, Weblate has notifications built in, but it seems that nobody subscribes to them.

> I think that (at least for somewhat experienced translators) "raw" editing of the po files is the best way to edit translations - be it in a text editor or via a GUI app like Poedit or Gtranslator.

One missing feature here, compared to Weblate, is screenshots. Is there a way to annotate .po files with links to screenshots? I also like how it can automatically recognize strings on the screenshot (of course, that can be achieved with a script that runs some OCR tool...)

> However, I see that translating software is a good way for not-so-tech-savy people to contribute and give back to the FOSS community. Hence, the barrier of entry should be quite low. Learning git, GitHub etc. might already be too much of a hassle...
What do you think?

So far more people contributed through Weblate than through GitHub. But I'm not sure if it's a good thing - people translate some part of the UI, then disappear, and I'm stuck with half-translated interface.

**Comment by daPhipz** [comment]
> Weblate creates a pull request but, unfortunately, only one for all languages. It squashes all commits by author, though, so I can cherry-pick them...

Is there some form of proofreading going on when using Weblate? For example, is a string shown to multiple translators before it gets in a PR to merge into master? I think that (where possible), there should be some sort of review process going on (see the [latest news](https://www.omgubuntu.co.uk/2023/10/ubuntu-23-10-installer-translations-hijack) on the Ubuntu 23.10 release, for example).

> Hm. So imagine I'm adding a new UI element. What should I do? Asking all translators to translate it immediately, before merging the PR, is not an option, I think.

Oh, I am sorry - I got this wrong! I thought you were wondering how we make sure that the strings are marked as translatable xD
You could implement a string freeze period before a release to give the translators a dedicated timeframe where no new strings are added, and no strings are changed. I think if you file an issue minimum a week before a new release, and ping all translators in it, there should be enough time to translate all strings that need an update. Translating every new string right when it gets merged into master is a waste of time IMO, since it might get changed again before the release.

> Is there a way to annotate .po files with links to screenshots?

No. You can however use [comments](https://wiki.gnome.org/TranslationProject/DevGuidelines/Use%20comments) and [translation contexts](https://wiki.gnome.org/TranslationProject/DevGuidelines/Translation%20contexts) in order to give some context to the translators. Of course, you could link to a screenshot in a comment, if you absolutely need to.

**Reply by amezin** [reply]
> Is there some form of proofreading going on when using Weblate? For example, is a string shown to multiple translators before it gets in a PR to merge into master? I think that (where possible), there should be some sort of review process going on (see the [latest news](https://www.omgubuntu.co.uk/2023/10/ubuntu-23-10-installer-translations-hijack) on the Ubuntu 23.10 release, for example).

On Weblate, strings can be marked as "approved". Then people outside of the project could only make suggestions for those strings, and members of the project can accept or decline the changes. I've marked ru and fr translations as "approved" initially. But the problem is - you need someone to actually go and review the suggestions from time to time. Which isn't happening with the French translation, as far as I can tell...

Also, usually I take a look at what I'm merging, if it's from someone I see for the first time.

**Reply by amezin** [reply]
If you or @DasJott registered on Weblate and accepted invitation to ddterm project, we could also mark de translation "approved" - and activate suggestions/reviews for it

BTW

@DasJott I thought you were using Weblate and I thought I've invited you to the project there already. But I can't find you in the list.

**Reply by DasJott** [reply]
@amezin I am using Weblate and you invited me, right! I thought I already accepted the invitation? I will check.

**Reply by DasJott** [reply]
I accepted the invitation from Github. Just saw I got invited to Weblate as well, but not sure if I also accepted that one.
If I click on the link in the according email (must have overseen this one) I get a 404, so possibly it needs to be resent?
Sorry for the inconvenience... this time of the year is full of birthdays and I have only little time to spend on the computer :D

**Reply by DasJott** [reply]
There are strings to be reviewed but I can not approve those, yet.

**Reply by amezin** [reply]
Sent the invitation again, please check your email. After joining the project there, you should be able to approve/review. 

If you mark the translation as "approved" (thus requiring reviews for changes from outside collaborators), please also subscribe to notifications, and check if there are new suggestions to review from time to time.

**Comment by amezin** [comment]
@ddterm/translators Does anybody know, are there any alternatives to Weblate: hosted, free for OSS, making git commits with correct authorship? Cause I'm a bit tired of Weblate's issues
